crates/subspace-core-primitives/src/lib.rs

```diff
@@ -95,23 +93,122 @@ pub type Blake3Hash = [u8; BLAKE3_HASH_SIZE];
     Hash,
     From,
     Into,
+    AsRef,
+    AsMut,
     Deref,
     DerefMut,
     Encode,
     Decode,
     TypeInfo,
     MaxEncodedLen,
 )]
+#[into(ref, ref_mut)]
+#[as_ref(forward)]
+#[as_mut(forward)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "serde", serde(transparent))]
-pub struct Blake3HashHex(#[cfg_attr(feature = "serde", serde(with = "hex"))] Blake3Hash);
+pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; BLAKE3_HASH_SIZE]);
```

Comment:
We should have a constant with size in the type instead of standalone and remove `BLAKE3_HASH_SIZE`:
```suggestion
pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
```

Reply:
This works except for using it in the type definition itself, which causes a compile cycle error. The workaround is to use `32` in the type definition and `Blake3Hash::SIZE` everywhere else.

<details>
<summary>full error</summary>

```
error[E0391]: cycle detected when evaluating type-level constant
   --> crates/subspace-core-primitives/src/lib.rs:104:80
    |
104 | pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
    |                                                                                ^^^^^^^^^^
    |
note: ...which requires const-evaluating + checking `<impl at crates/subspace-core-primitives/src/lib.rs:92:5: 92:9>::{constant#0}`...
   --> crates/subspace-core-primitives/src/lib.rs:104:80
    |
104 | pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
    |                                                                                ^^^^^^^^^^
note: ...which requires caching mir of `<impl at crates/subspace-core-primitives/src/lib.rs:92:5: 92:9>::{constant#0}` for CTFE...
   --> crates/subspace-core-primitives/src/lib.rs:104:80
    |
104 | pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
    |                                                                                ^^^^^^^^^^
note: ...which requires elaborating drops for `<impl at crates/subspace-core-primitives/src/lib.rs:92:5: 92:9>::{constant#0}`...
   --> crates/subspace-core-primitives/src/lib.rs:104:80
    |
104 | pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
    |                                                                                ^^^^^^^^^^
note: ...which requires borrow-checking `<impl at crates/subspace-core-primitives/src/lib.rs:92:5: 92:9>::{constant#0}`...
   --> crates/subspace-core-primitives/src/lib.rs:104:80
    |
104 | pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
    |                                                                                ^^^^^^^^^^
note: ...which requires promoting constants in MIR for `<impl at crates/subspace-core-primitives/src/lib.rs:92:5: 92:9>::{constant#0}`...
   --> crates/subspace-core-primitives/src/lib.rs:104:80
    |
104 | pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
    |                                                                                ^^^^^^^^^^
note: ...which requires const checking `<impl at crates/subspace-core-primitives/src/lib.rs:92:5: 92:9>::{constant#0}`...
   --> crates/subspace-core-primitives/src/lib.rs:104:80
    |
104 | pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
    |                                                                                ^^^^^^^^^^
note: ...which requires building MIR for `<impl at crates/subspace-core-primitives/src/lib.rs:92:5: 92:9>::{constant#0}`...
   --> crates/subspace-core-primitives/src/lib.rs:104:80
    |
104 | pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
    |                                                                                ^^^^^^^^^^
note: ...which requires match-checking `<impl at crates/subspace-core-primitives/src/lib.rs:92:5: 92:9>::{constant#0}`...
   --> crates/subspace-core-primitives/src/lib.rs:104:80
    |
104 | pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
    |                                                                                ^^^^^^^^^^
note: ...which requires type-checking `<impl at crates/subspace-core-primitives/src/lib.rs:92:5: 92:9>::{constant#0}`...
   --> crates/subspace-core-primitives/src/lib.rs:104:80
    |
104 | pub struct Blake3Hash(#[cfg_attr(feature = "serde", serde(with = "hex"))] [u8; Self::SIZE]);
    |                                                                                ^^^^^^^^^^
    = note: ...which again requires evaluating type-level constant, completing the cycle
note: cycle used when building specialization graph of trait `core::convert::From`
   --> /Users/hyper/.rustup/toolchains/nightly-2024-08-01-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:580:1
    |
580 | pub trait From<T>: Sized {
    | ^^^^^^^^^^^^^^^^^^^^^^^^
    = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
```

</details>

Reply:
You just need to implement one of the derives (`Into`) manually, see https://github.com/JelteF/derive_more/issues/413